### PR TITLE
Fix: Hvdc active power setpoint should be interpreted on rectifier AC side

### PR DESCRIPTION
**Pull request by annetill**
Signed-off-by: Anne Tilloy <anne.tilloy@rte-france.com>

**Please check if the PR fulfills these requirements** *(please use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes)*
- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**Does this PR already have an issue describing the problem ?** *If so, link to this issue using `'#XXX'` and skip the rest*

No.

**What kind of change does this PR introduce?** *(Bug fix, feature, docs update, ...)*

The Hvdc active power set point is applied on converter station that is in rectifier mode, on its AC side. On the inverter AC side, we computes the active power using both loss factors and the losses due to the line itself.

**What is the current behavior?** *(You can also link to an open issue here)*

The Hvdc active power set point is applied on converter station that is in rectifier mode, on its DC side.

**What is the new behavior (if this is a feature change)?**

**Does this PR introduce a breaking change or deprecate an API?** *If yes, check the following:*
- [ ] The *Breaking Change* or *Deprecated* label has been added
- [ ] The migration guide has been updated in the github wiki *(What changes might users need to make in their application due to this PR?)*

**Other information**:

(if any of the questions/checkboxes don't apply, please delete them entirely)

### CODE REVIEW

**Review by sylvlecl — APPROVED** [review]
Minor comments about doc and implementation detail

File: src/main/java/com/powsybl/openloadflow/network/impl/HvdcConverterStations.java [hunk 1]
```diff
@@ -39,6 +40,86 @@ public static double getActivePowerSetpointMultiplier(HvdcConverterStation<?> st
...
+        }
+    }
+
+    private static double getHvdcLineLosses(double rectifierPDc, double nominalV, double r) {
+        // This method computes the losses due to the HVDC line.
+        // The active power value on rectifier DC side is known as the HVDC active power set point minus the losses related
+        // to AC/DC conversion (rectifier conversion), the voltage is approximated to the nominal voltage as attribute of the HVDC line.
+        // In an HVDC, as a branch with two sides, the difference between pDc1 and pDc2 can be computed with the assumptions:
+        // I = (V1 - V2) / R and pDc1 = I * V1 and pDc2 = I * V2
+        // we simply obtain that the absolute value of the difference is equal to R * pDc1 * pDc1 / (V1 * V1) if side 1 is rectifier side.
```
[the remark this answers is not in the document]
Comment (annetill) [reply src/main/java/com/powsybl/openloadflow/network/impl/HvdcConverterStations.java:93]: I have checked and Dynawo is making the same assumption.

File: src/main/java/com/powsybl/openloadflow/network/impl/HvdcConverterStations.java [hunk 2]
```diff
@@ -29,7 +29,8 @@ public static boolean isRectifier(HvdcConverterStation<?> station) {
                 || (line.getConverterStation2() == station && line.getConvertersMode() == HvdcLine.ConvertersMode.SIDE_1_INVERTER_SIDE_2_RECTIFIER);
     }
 
-    public static double getActivePowerSetpointMultiplier(HvdcConverterStation<?> station) {
+    public static double getSign(HvdcConverterStation<?> station) {
```
[the remark this answers is not in the document]
Comment (annetill) [reply src/main/java/com/powsybl/openloadflow/network/impl/HvdcConverterStations.java:32]: Yes, but I prefer to put everything in the same class, to explicit totally what we are doing with the signs... for DC sensitivity analysis, it is really a part for me.

File: src/main/java/com/powsybl/openloadflow/network/impl/HvdcConverterStations.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (annetill) [reply src/main/java/com/powsybl/openloadflow/network/impl/HvdcConverterStations.java:93]: In Dynawo:
`void
ModelHvdcLink::setConvertersActivePower(const shared_ptr<HvdcLineInterface>& dcLine) {
  if (dcLine->getConverter1()->hasP() && dcLine->getConverter2()->hasP()) {
    // retrieve active power at the two points of common coupling from load flow data in IIDM file
    P01_ = -dcLine->getConverter1()->getP() / SNREF;
    P02_ = -dcLine->getConverter2()->getP() / SNREF;
  } else {
    // calculate losses on dc line
    double PdcLoss = rdc_ * (pSetPoint_ / vdcNom_) * (pSetPoint_ / vdcNom_) / SNREF;  // in pu

    // calculate active power at the two points of common coupling (generator convention)
    double P0dc = pSetPoint_ / SNREF;  // in pu
    if (converterMode_ == HvdcLineInterface::RECTIFIER_INVERTER) {
      P01_ = -P0dc;  // RECTIFIER (absorbs power from the grid)
      P02_ = ((P0dc * (1 - lossFactor1_)) - PdcLoss) * (1. - lossFactor2_);  // INVERTER (injects power to the grid)
    } else {   // converterMode_ == HvdcLineInterface::INVERTER_RECTIFIER
      P01_ = ((P0dc * (1 - lossFactor2_)) - PdcLoss) * (1. - lossFactor1_);  // INVERTER (injects power to the grid)
      P02_ = -P0dc;  // RECTIFIER (absorbs power from the grid)
    }
  }
}`

[the remark this answers is not in the document]
Comment (sylvlecl) [reply src/main/java/com/powsybl/openloadflow/network/impl/HvdcConverterStations.java:93]: Actually it's even more approximated: the sepoint on AC side is used to compute the losses ?

File: src/main/java/com/powsybl/openloadflow/network/impl/HvdcConverterStations.java [hunk 3]
```diff
@@ -39,6 +40,86 @@ public static double getActivePowerSetpointMultiplier(HvdcConverterStation<?> st
...
+
+    private static double getAbsoluteValueInverterPAc(double rectifierLossFactor, double inverterLossFactor,
+                                                      HvdcLine hvdcLine) {
+        // On inverter side, absolute value of PAc of a VSC converter station should be computed in three step:
+        // 1) compute the losses related to the rectifier conversion.
+        // 2) compute the losses related to the HVDC line itself (R i^2).
+        // 3) compute the losses related to the inverter conversion.
+        double rectifierPDc = hvdcLine.getActivePowerSetpoint() * (1 - rectifierLossFactor / 100); // rectifierPDc positive.
+        double inverterPDc = rectifierPDc - getHvdcLineLosses(rectifierPDc, hvdcLine.getNominalV(), hvdcLine.getR());
+        return inverterPDc * (1 - inverterLossFactor / 100); // always positive.
```
[the remark this answers is not in the document]
Comment (AnkurArohi) [reply src/main/java/com/powsybl/openloadflow/network/impl/HvdcConverterStations.java:105]: Yes this is an important information is it also like that for OLF module?

[the remark this answers is not in the document]
Comment (annetill) [reply src/main/java/com/powsybl/openloadflow/network/impl/HvdcConverterStations.java:105]: I am going to increase the javadoc, but for my defense it is in the web-site ;-) https://www.powsybl.org/pages/documentation/grid/model/

**Pull request merged**